ThunkLibs/include/common/Guest.h

```diff
@@ -101,10 +101,30 @@ inline bool IsLibLoaded(const char *libname) {
 template<auto Thunk, typename Result, typename... Args>
 inline Result CallHostFunction(Args... args) {
 #ifndef _M_ARM_64
-    uintptr_t host_addr;
-    asm volatile("mov %%r11, %0" : "=r" (host_addr));
+  // This magic incantation of using a register variable with an empty asm block is necessary for correct operation!
+  // If we only use inline asm that sets a variable through inline asm then the compiler will reorder the function
+  // prologue to be BEFORE our inline asm. Which makes sense in hindsight, but for anything with 8+ arguments then
+  // will clobber our r11 register we save the data that is inside of it.
+
+  // First we need to declare the r11 register variable
+  register uintptr_t host_addr asm ("r11");
+
+  // We then create an empty *volatile* asm block saying that it is assigning the register variable.
+  // Yes, it is already set coming in to this function due to custom ABI.
+  // This gets both GCC and Clang to understand that the variable is set, seemingly at the start of the function.
+  // So its own internal live-range tracking extends its begining range to the start of the function.
+  //
+  // To verify this in the future, search for `mov     r11` in binaryninja, and ensure that all uses inside of `CallHostFunction`
+  // don't have intersecting ranges.
+  //
+  // Specifically when clang is used to compile thunks, it is significantly more aggressive at using R11, so it is more likely to use it.
```

Comment:
```suggestion
  // Note that this issue is more likely to occur when clang is used to compile thunks, since its optimizer is more aggressive at using R11.
```